core/lib/zksync_core/src/fee_model.rs

```diff
@@ -0,0 +1,143 @@
+            MainNodeFeeParams::V1(params) => {
+                compute_batch_fee_model_input_v1(params, l1_gas_price_scale_factor)
+            }
+            MainNodeFeeParams::V2(params) => compute_batch_fee_model_input_v2(
+                params,
+                l1_gas_price_scale_factor,
+                l1_pubdata_price_scale_factor,
+            ),
+        }
+    }
+
+    /// Returns the batch fee input as-is, i.e. without any scaling for the L1 gas and pubdata prices.
+    fn get_batch_fee_input(&self) -> BatchFeeInput {
+        self.get_batch_fee_input_scaled(1.0, 1.0)
+    }
+
+    /// Returns the fee model parameters.
+    fn get_fee_model_params(&self) -> MainNodeFeeParams;
+}
+
+/// The struct that represents the batch fee input provider to be used in the main node of the server, i.e.
+/// it explicitly gets the L1 gas price from the provider and uses it to calculate the batch fee input instead of getting
+/// it from other node.
+#[derive(Debug)]
+pub(crate) struct MainNodeFeeInputProvider<G: ?Sized> {
+    provider: Arc<G>,
+    config: MainNodeFeeModelConfig,
+}
+
+impl<G: L1GasPriceProvider + ?Sized> BatchFeeModelInputProvider for MainNodeFeeInputProvider<G> {
+    fn get_fee_model_params(&self) -> MainNodeFeeParams {
+        match self.config {
+            MainNodeFeeModelConfig::V1(config) => MainNodeFeeParams::V1(MainNodeFeeParamsV1 {
+                config,
+                l1_gas_price: self.provider.estimate_effective_gas_price(),
+            }),
+            MainNodeFeeModelConfig::V2(config) => MainNodeFeeParams::V2(MainNodeFeeParamsV2 {
+                config,
+                l1_gas_price: self.provider.estimate_effective_gas_price(),
+                l1_pubdata_price: self.provider.estimate_effective_pubdata_price(),
+            }),
+        }
+    }
+}
+
+impl<G: L1GasPriceProvider + ?Sized> MainNodeFeeInputProvider<G> {
+    pub(crate) fn new(provider: Arc<G>, config: MainNodeFeeModelConfig) -> Self {
+        Self { provider, config }
+    }
+}
+
+/// Calculates the batch fee input based on the main node parameters.
+/// This function uses the `V1` fee model, i.e. where the pubdata price does not include the proving costs.
+pub(crate) fn compute_batch_fee_model_input_v1(
+    params: MainNodeFeeParamsV1,
+    l1_gas_price_scale_factor: f64,
+) -> BatchFeeInput {
+    let l1_gas_price = (params.l1_gas_price as f64 * l1_gas_price_scale_factor) as u64;
+
+    BatchFeeInput::L1Pegged(L1PeggedBatchFeeModelInput {
+        l1_gas_price,
+        fair_l2_gas_price: params.config.minimal_l2_gas_price,
+    })
+}
+
+/// Calculates the batch fee input based on the main node parameters.
+/// This function uses the `V2` fee model, i.e. where the pubdata price does not include the proving costs.
+pub(crate) fn compute_batch_fee_model_input_v2(
+    params: MainNodeFeeParamsV2,
+    l1_gas_price_scale_factor: f64,
+    l1_pubdata_price_scale_factor: f64,
+) -> BatchFeeInput {
+    let MainNodeFeeParamsV2 {
```

Comment:
optional: should we check that scale factor is valid (for example > 0,  and < 100 etc?)

Reply:
it is mainly an optional param that is used in the API... I would personally not add unneeded constraints to it

Reply:
I'd still check that it is > 0..

if someone sets it to -1 --> the effects will be very strange (and potentially dangerous)